src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java

```diff
@@ -57,66 +74,169 @@
      *            The instance of our {@link SlimefunPlugin}
      */
     public void start(@Nonnull SlimefunPlugin plugin) {
+
         this.tickRate = SlimefunPlugin.getCfg().getInt("URID.custom-ticker-delay");
 
         BukkitScheduler scheduler = plugin.getServer().getScheduler();
-        scheduler.runTaskTimerAsynchronously(plugin, this, 100L, tickRate);
+        scheduler.runTaskLater(plugin, this, 100L);
     }
 
-    /**
-     * This method resets this {@link TickerTask} to run again.
-     */
-    public void reset() {
-        running = false;
+    @Override 
+    public final void run() {
+
+        // This should not happen. Only case where this may happen is if this task is scheduled or started multiple times.
+        if (running) {
+            return;
+        }
+        if (!Bukkit.isPrimaryThread()) {
+            return;
+        }
+
+        // Don't overload the Bukkit scheduler with a thousand runnables. It doesn't like that. This way, we can spread the
+        // time cost over several game ticks and adapt ticker rate to server performance.
+        final ArrayBlockingQueue<Runnable> syncTasks = new ArrayBlockingQueue<>(32);
+
+        // Iterate over the ticker tasks. Process the asynchronous ones on the Bukkit async task thread, and queue the sync
+        // tasks to the queue.
+        Bukkit.getScheduler().runTaskAsynchronously(
+                SlimefunPlugin.instance(),
+                () -> this.tick(syncTasks)
+        );
+
+        // Process the synchronous ticker tasks on the main thread.
+        processSyncTasks(syncTasks);
     }
 
-    @Override
-    public void run() {
+    private void processSyncTasks(final @Nonnull BlockingQueue<Runnable> tasks) {
```

Comment:
No need to mark this parameter as final really.

Reply:
It is needed. We use it in a lambda which requires it as final.

Reply:
Ah okay, can you at least put the annotation before the final then?